### PR TITLE
Clean up LuceneDocumentFromRecord

### LINKED ISSUE #1235: Lucene should support more field data types
At least numbers. I think we also want to differentiate between `StringField` and `TextField`, both of which will be strings in Protobuf. I'm not sure what the cleanest way to do that is, but one idea would have to have a "wrapping" key expression that just meant that: it wouldn't affect evaluation itself, but would be noted by the document from index entry builder.

### LINKED ISSUE #1297: Lucene indexing of map-like fields
The keys (which would need to be strings, probably with some character set restrictions) could become additional fields in the document, continuing the existing method of concatenating a path name from parent record.

As with #1235, this may argue for some kind(s) of wrapping key expression.

### PR DESCRIPTION
**Pull request by MMcM**
The main addition is a new class, `LuceneIndexExpressions`, which implements in a single place knowledge of what components are allowed in a Lucene index key expression and how they nest to inform document creation. This same class walks the tree either against `FDBRecord`s to make Lucene document fields or against Protobuf meta-data to symbolically represent what that _would_ do.

The old code had an implementation for the next two, but they were never closed:
* Resolves #1235: Lucene should support more field data types
* Resolves #1297: Lucene indexing of map-like fields

This new scheme bases the mapping on explicit function calls, rather than heuristics.
The default index on a concatenation of fields will have single tokens, not full text.

This only minimally addresses the Lucene planner: just enough to keep the tests running.

This eliminates the need for special Lucene key expressions (other than the functions), which in turn eliminates the need for special serialization post-processing. That entire mechanism is removed (a core change).

### CODE REVIEW

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePlanner.java [hunk 1]
```diff
@@ -229,9 +229,11 @@ private LuceneIndexQueryPlan getComparisonsForLuceneFilter(@Nonnull Index index,
         return null;
     }
 
-    public boolean validateIndexField(@Nonnull Index index,
-                                       @Nonnull String field) {
-        return listIndexFieldNames(index.getRootExpression()).contains(field);
+    public boolean validateIndexField(@Nonnull Index index, @Nonnull String field) {
+        final RecordType recordType = getRecordMetaData().recordTypesForIndex(index).iterator().next();
```
Comment (tian-yizuo) [inline fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePlanner.java:233]: Looks this is being used in multiple places. So does it make sense to have this as a method in ```RecordMetaData``` to be reused?

**Review by tian-yizuo — COMMENTED** [review]
This looks good to me

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePlanner.java (hunk 1 above)
Comment (MMcM) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LucenePlanner.java]: I changed these to check all record types, so they are no longer identical.

**Review by ohadzeliger — COMMENTED** [review]
No issues - just a few questions

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexExpressions.java [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    public static void validate(@Nonnull KeyExpression root, @Nonnull Descriptors.Descriptor recordType) {
+        getFields(root, new MetaDataSource(recordType), (fieldName, value, type, stored) -> {
+        });
+    }
+
+    /**
+     * Get the types of known document fields.
+     * @param root the {@code LUCENE} index root expresison
+     * @param recordType Protobuf meta-data for record type
+     * @return a list of document field names
```
[the remark this answers is not in the document]
Comment (MMcM) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexExpressions.java]: Fixed.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexExpressions.java [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+
+        Iterable<RecordSource> getChildren(FieldKeyExpression parentExpression);
+
+        Iterable<Object> getValues(FieldKeyExpression fieldExpression);
+    }
+
+    /**
+     * An actual document / document meta-data.
+     */
+    public interface DocumentDestination {
```
[the remark this answers is not in the document]
Comment (MMcM) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexExpressions.java:124]: I do not think so. There are document fields and there are record fields and field alone is ambiguous.

Perhaps `Source` and `Destination` are not helpful. But the overall context is converting records with record fields into documents with document fields.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexExpressions.java [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+                    fieldNameSuffix = (String)names.next();
+                    if (names.hasNext()) {
+                        throw new RecordCoreException("Lucene field name override should evaluate to single value");
+                    }
+                }
+            } else {
+                throw new RecordCoreException("Lucene field name override should be a literal or a field");
+            }
+            suffixOverride = true;
+            expression = fieldNameExpression.getNamedExpression();
```
[the remark this answers is not in the document]
Comment (MMcM) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexExpressions.java:176]: This is the named expression -- the thing being given a name, not the naming expression -- that name itself.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexExpressions.java [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+            }
+            String fieldName = appendFieldName(fieldNamePrefix, fieldNameSuffix);
+            for (RecordSource subsource : source.getChildren(parentExpression)) {
+                getFields(child, subsource, destination, fieldName);
+            }
+            return;
+        }
+
+        boolean fieldStored = false;
+        if (expression instanceof LuceneFunctionKeyExpression.LuceneStored) {
```
[the remark this answers is not in the document]
Comment (MMcM) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexExpressions.java]: Sure, there is no reason not to allow that. Fixed.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneFunctionKeyExpression.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        public List<Descriptors.FieldDescriptor> validate(@Nonnull Descriptors.Descriptor descriptor) {
+            final List<Descriptors.FieldDescriptor> result = super.validate(descriptor);
+            if (!(arguments instanceof ThenKeyExpression && ((ThenKeyExpression)arguments).getChildren().size() == 2 && ((ThenKeyExpression)arguments).getChildren().get(1).getColumnSize() == 1)) {
+                throw new InvalidExpressionException("Lucene field name subexpression should be single column");
+            }
+            return result;
+        }
+
+        @Nonnull
+        public KeyExpression getNamedExpression() {
```
[the remark this answers is not in the document]
Comment (MMcM) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneFunctionKeyExpression.java:107]: Added comments.

File: fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexExpressions.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ohadzeliger) [reply fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneIndexExpressions.java:124]: Makes sense. Thanks.

**Review by ohadzeliger — DISMISSED** [review]

**Review by ohadzeliger — DISMISSED** [review]

**Review by tian-yizuo — DISMISSED** [review]

**Review by tian-yizuo — APPROVED** [review]

**Pull request merged**